### DISCUSSION TITLE
Experiences UI testing the Uno Platform Android app with Appium

### DISCUSSION THREAD

**Discussion by jamesmcroft** [unoplatform/uno#6225] [Q&A]
Quick question, has anyone had any luck running UI tests against an Uno Platform application running on Android using Appium? 

I'm currently struggling to get the driver to find any elements in the UI. When I use the layout inspector in Android Studio against the application, I can see the content description containing the automation properties that I'm trying to access. 

Knowing how you should access UI elements using the AndroidUIAutomator, I've constructed this quite simple select extension for elements by automation ID but it fails every time. 

```csharp
androidDriver.FindElementByAndroidUIAutomator(new AndroidUiSelector().DescriptionEquals(automationId))
```

This extension for the framework is based on a similar extension for native UI elements that I created in another UI test library. 

Note, I am able to get the application up and running with automation in Appium, the problem is just locating items by the XAML automation ID.

There is a [sample UI test project in the Legerity framework](https://github.com/MADE-Apps/legerity-uno/tree/main/samples/UnoSampleAppTests) that has a simple button click test under `Tests/ButtonTests` (currently disabled for Android due to this issue).

**Comment by jeromelaban** [comment]
We have not spent time working with Appium at this point. If you set the same configuration properties as in Uno.UITest documentation, the description field should be filled properly, though.

**Reply by jamesmcroft** [reply]
No worries at all on the Appium front. That's what I'm trying to resolve and I'm nearly there. 

I will double check that everything is setup correctly. Seeing the content description populated but it's not finding the elements.

**Reply by jamesmcroft** [reply]
Think I may have worked it out and it's a quirk of Appium and Android. Having not seen this before, I thought it could be the way Uno was generating the UI for Android. 

Essentially, it looks like with Appium, the automation ID is being populated in the TagName property of the Appium element. When requesting it as content description, it returns null.

**Reply by jeromelaban** [reply]
Could it be that appium capture the UI hierarchy before the automationId are set? Xamarin.UITest has a similar quirk in some conditions. In other cases, it could be related to out-of-viewport elements.

**Reply by jamesmcroft** [reply]
This could be the case. What I've been seeing when finding all elements is that some have their content description set where others aren't. Depending on how complex the page is that is.

**Reply by jamesmcroft** [reply]
After doing an export of what Appium can see, it looks like there are a couple of problems, I believe are caused by Android.

For some controls, the content description isn't being set. However, there are examples like the `TextBlock`, `CheckBox`, or `Button` that set the content description as their content text rather than the automation property. 

I'm still having a play around with this and I'll see what I can produce. Thanks for the pointers @jeromelaban

**Reply by jeromelaban** [reply]
In this case, it's generally caused by the fact that automation "mode" is not enabled properly. For TextBlock, this is done here:
https://github.com/unoplatform/uno/blob/33543f670336f62713d9b610031d0b792e65f5c8/src/Uno.UI/UI/Xaml/Controls/TextBlock/TextBlock.Android.cs#L122

You can try setting the `FrameworkElementHelper.IsUiAutomationMappingEnabled` property explicitly.

**Reply by jamesmcroft** [reply]
I might have a play with this inside the Uno repository because for some reason it doesn't appear to be working this way with the sample applications I've built with that flag on. This includes if you set the `IsUiAutomationMappingEnabled` property in the csproj or the `FrameworkElementHelper.IsUiAutomationMappingEnabled` property in the App.xaml.cs.

**Reply by jamesmcroft** [reply]
Looks like it works fine in Uno. I believe there might be a bug in Appium.
